### PR TITLE
grocery-store jessie zhang

### PR DESCRIPTION
**Pull request by jessiezhang2017**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | It  will help to prevent the methods are called with invalid arguments.The ruby will provide stack trace to help the programers to debug.
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? |The .all and .find methods need to calling on Class, since the instance methods can only work on one instance, but the .all and .find are dealing with all the orders have been  created , therefore should deal with more than one instances. 
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? |the relation ship between Order and Customer are one to many.  One customer can make many orders.  one order only has one customer. In Solar System, only Solar system can have many planets.  only planet only belong to  one Solar system. 
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | in CSV file, each record of the Order contains one customer ID.   In my program, I create an instance of Order class, which use customer ID as the instance variable. we can build up the program different than the original file set ups, if the program can store data , do the searches and calculations efficiently and correctly. 
Did the presence of automated tests change the way you thought about the problem? How? |Yes, it help me to think more about the edge cases which I might ignore before. 
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? | O(C) for Customer.find .  O(R) for Order.find

### CODE REVIEW

File: lib/order.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  # create a method to calculate the sum after tax
+  def total
+    sum_before_tax = self.products.values.sum
+    sum_after_tax = sum_before_tax * 1.075
+    return sum_after_tax.round(2)
+  end
+  # a method to add product
+  def add_product(prd_name, prd_price)
+    if @products.keys.include? (prd_name)
+      raise ArgumentError,"Product namne exists, try other nanmes"
```
Comment (CheezItMan) [inline lib/order.rb:29]: spelling 😊

File: lib/order.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
+require 'pry'
+require_relative 'customer'
+
+class Order
+  attr_reader :id, :customer, :products, :fulfillment_status
```
Comment (CheezItMan) [inline lib/order.rb:5]: The customer, products and fullfillment_status should be accessors not readers.

File: lib/order.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      new_order = self.new(order_id, product_h, customer, fulfillment_status)
+      all_orders.push(new_order)
+    end
+
+    return all_orders
+  end
+  # find an order with the given order ID
+  def self.find(id)
+    all_orders = self.all
+    order_found = all_orders.select {|order| order.id == id}
```
Comment (CheezItMan) [inline lib/order.rb:73]: `find` will work just as well.

File: specs/order_spec.rb [hunk 4]
```diff
@@ -111,13 +112,52 @@
...
+
+    it "Is removed from the collection of products" do
+      products = { "banana" => 1.99, "cracker" => 3.00 }
+      order = Order.new(1337, products, customer)
+
+      order.remove_product("cracker")
+      expect(order.products.include?("cracker")).must_equal false
+    end
+
+    it "Raises an ArgumentError if no product was found" do
```
Comment (CheezItMan) [inline specs/order_spec.rb:135]: Good tests!

File: specs/order_spec.rb [hunk 5]
```diff
@@ -139,23 +179,70 @@
...
+        "Amaranth" => 83.81,
+        "Smoked Trout" => 70.6,
+        "Cheddar" => 5.63
+      }
+      customer_id = 20
+      fulfillment_status = :pending
+
+      order = Order.all.last
+
+      # Check that all data was loaded as expected
```
Comment (CheezItMan) [inline specs/order_spec.rb:195]: 👍

**Comment by CheezItMan** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | Check
Used Git Regularly | More commits would be better and avoid using the waves in the commit messages and instead describe the changes.  
**Wave 1** |
All provided tests pass | Check
Using the appropriate `attr_` for instance variables | Check
**Wave 2** |
All stubbed tests are implemented fully and pass | Check
Used `CSV` library only in `.all` (not in `.find`) | Check
Appropriately parses the product data from CSV file in `Order.all` | Check
`Order.all` calls `Customer.find` to set up the composition relation | Check
orders_by_customer | Nice work on this extra
**Additional Notes** | Great job.  You hit all the learning goals.
